packages/block-library/src/columns/edit.js

```diff
@@ -0,0 +1,121 @@
+/**
+ * External dependencies
+ */
+import classnames from 'classnames';
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { compose } from '@wordpress/compose';
+import {
+	PanelBody,
+	RangeControl,
+} from '@wordpress/components';
+import { Fragment } from '@wordpress/element';
+import {
+	InspectorControls,
+	InnerBlocks,
+	BlockControls,
+	BlockVerticalAlignmentToolbar,
+} from '@wordpress/block-editor';
+import { withSelect, withDispatch } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { getColumnsTemplate } from './utils';
+
+/**
+ * Allowed blocks constant is passed to InnerBlocks precisely as specified here.
+ * The contents of the array should never change.
+ * The array should contain the name of each block that is allowed.
+ * In columns block, the only block we allow is 'core/column'.
+ *
+ * @constant
+ * @type {string[]}
+*/
+const ALLOWED_BLOCKS = [ 'core/column' ];
+
+export const ColumnsEdit = function( { attributes, setAttributes, className, updateAlignment } ) {
+	const { columns, verticalAlignment } = attributes;
+
+	const classes = classnames( className, `has-${ columns }-columns`, {
+		[ `are-vertically-aligned-${ verticalAlignment }` ]: verticalAlignment,
+	} );
+
+	const onChange = ( alignment ) => {
+		// Update all the (immediate) child Column Blocks
+		updateAlignment( alignment );
+	};
+
+	return (
+		<Fragment>
+			<InspectorControls>
+				<PanelBody>
+					<RangeControl
+						label={ __( 'Columns' ) }
+						value={ columns }
+						onChange={ ( nextColumns ) => {
+							setAttributes( {
+								columns: nextColumns,
+							} );
+						} }
+						min={ 2 }
+						max={ 6 }
+					/>
+				</PanelBody>
+			</InspectorControls>
+			<BlockControls>
+				<BlockVerticalAlignmentToolbar
+					onChange={ onChange }
+					value={ verticalAlignment }
+				/>
+			</BlockControls>
+			<div className={ classes }>
+				<InnerBlocks
+					template={ getColumnsTemplate( columns ) }
+					templateLock="all"
+					allowedBlocks={ ALLOWED_BLOCKS } />
+			</div>
+		</Fragment>
+	);
+};
+
+export default compose(
+	/**
+	 * Selects the child column Blocks for this parent Column
+	 */
+	withSelect( ( select, { clientId } ) => {
+		const { getBlocksByClientId } = select( 'core/editor' );
+
+		return {
+			childColumns: getBlocksByClientId( clientId )[ 0 ].innerBlocks,
```

Comment:
There's a few levels of indirection here to get to the `innerBlocks`. You could have called `getBlock` to get the singular block by `clientId`, but actually think it may have been better to call:

```js
getBlocks( clientId )
```

...where the `clientId` argument of the `getBlocks` selector signifies the "root" (i.e. parent) from which to select inner blocks.

https://wordpress.org/gutenberg/handbook/designers-developers/developers/data/data-core-block-editor/#getblocks